src/App.tsx

```diff
@@ -1,156 +1,237 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+/* eslint-disable no-param-reassign */
+import classNames from 'classnames';
+import React, { useCallback, useMemo, useState } from 'react';
+
+type Todo = {
+  id: number;
+  title: string;
+  isEditing: boolean;
+};
+
+type Filter = 'All' | 'Completed' | 'Active';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [completedTodos, setCompletedTodos] = useState<Todo[]>([]);
+  const [inputValue, setInputValue] = useState('');
+  const [filter, setFilter] = useState<Filter>('All');
+  const [isVisible, setIsVisible] = useState(true);
+  const [isEditing, setIsEditing] = useState(false);
+  const [editedValue, setEditedValue] = useState('');
+
+  const createTodo = useCallback((title: string) => {
+    return {
+      id: Math.round(Math.random() * 1000) / 1000,
+      title,
+      isEditing: false,
+    } as Todo;
+  }, []);
+
+  const formSubmitHandler = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (inputValue.trim()) {
+      setTodos([...todos, createTodo(inputValue)]);
+      setInputValue('');
+    }
+  };
+
+  const editHandler = (event: React.FormEvent, todo: Todo) => {
+    event.preventDefault();
+
+    if (!editedValue.trim()) {
+      setTodos(todos.filter(currentTodo => currentTodo.id !== todo.id));
+    } else {
+      todo.title = editedValue;
+      todo.isEditing = false;
+    }
+
+    setIsEditing(false);
+    setEditedValue('');
+  };
+
+  const visibleTodos: Todo[] = useMemo(() => {
+    if (filter !== 'All') {
+      if (filter === 'Active') {
+        return [...todos.filter(todo => !completedTodos.includes(todo))];
+      } else {
+        return [...completedTodos];
+      }
+    } else {
+      return [...todos];
+    }
+  }, [todos, filter, completedTodos, isEditing]);
+
   return (
     <div className="todoapp">
       <h1 className="todoapp__title">todos</h1>
 
       <div className="todoapp__content">
         <header className="todoapp__header">
```

Comment:
split the logic and JSX between different components (at least Header, Footer, TodoList. But you can additionally create Filter, TodoItem, etc. components)